hardware-security-module/application-typescript/src/hsm-sample.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright IBM Corp. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as grpc from '@grpc/grpc-js';
+import * as crypto from 'crypto';
+import { connect, Gateway, HSMSigner, HSMSignerFactory, HSMSignerOptions, signers } from '@hyperledger/fabric-gateway';
+import * as fs from 'fs';
+import * as jsrsa from 'jsrsasign';
+import * as path from 'path';
+import { TextDecoder } from 'util';
+
+const mspId = 'Org1MSP';
+const user = 'HSMUser';
+const assetId = `asset${Date.now()}`;
+const utf8Decoder = new TextDecoder();
+
+// Sample uses fabric-ca-client generated HSM identities, certificate is located in the signcerts directory
+// and has been stored in a directory of the name given to the identity.
+
+const certPath = path.resolve(__dirname, '..', '..', 'crypto-material', 'hsm', user, 'signcerts', 'cert.pem');
+
+const tlsCertPath = path.resolve('..', '..', 'test-network','organizations','peerOrganizations', 'org1.example.com', 'peers', 'peer0.org1.example.com', 'tls', 'ca.crt');
+const peerEndpoint = 'localhost:7051';
+
+async function main() {
+    console.log('\nRunning the Node HSM sample');
+
+    // The gRPC client connection should be shared by all Gateway connections to this endpoint
+    const client = await newGrpcConnection();
+
+    // get an HSMSigner Factory. You only need to do this once for the application
+    const hsmSignerFactory = signers.newHSMSignerFactory(findSoftHSMPKCS11Lib());
+    const credentials = await fs.promises.readFile(certPath);
+
+    // Get the signer function and a close function. The close function closes the signer
+    // once there is no further need for it.
+    const {signer, close} = await newHSMSigner(hsmSignerFactory, credentials.toString());
+
+    const gateway = connect({
+        client,
+        identity: {mspId, credentials},
+        signer,
+    });
+
+    try {
+        await exampleTransaction(gateway);
+        console.log();
+        console.log('Node HSM sample completed successfully');
+    } finally {
```

Comment:
The flow here is a little bit suspect. If a failure happens before the `try` block, some resources may not get closed. You _could_ specify the variables that may need closing up-front (using `let`), then start the `try` block before you initialise them, then in the `finally` block use `.?` operator to only close/dispose things that got created before any failure occurred.

This is more correct but for the purposes of the sample probably doesn't matter too much as generally things will work.